### PR TITLE
Update Nikon flash tags

### LINKED ISSUE #1941: Add a decoder for Exif.Nikon3.FlashInfo
I received the following email:

_Hi Mr. Mills,_

_Some time ago, I contacted a developer of your cousin project exiftool, because I had managed to decode some private Nikon data when using a radio flash trigger, showing some settings for each group. This person completed my findings, asking for images with certain combinations of settings, and added the result to their project._

_In the meanwhile, I'm using only darktable which is based on exiv2, where this information is not available._

_Is there any possibility to get Exif.Nikon3.FlashInfo also decoded in exiv2?_

_Thanks_

_Christoph_

--------------

Here is my reply:

Christoph

Thank You for your email.  Exiv2 has a good and friendly working relationship with Phil Harvey.  An excellent engineer and very nice person.

In June, I retired from Exiv2 and will leave it to others to support and develop Exiv2 in future.  I volunteered to be the release engineer for v0.27.5 which is scheduled for release on 2021-10-22.  I released Exiv2 v0.27.5 RC3 today.  So, I’m almost out of the project.

I recommend that you open a new issue on https://github.com/exiv2/exiv2 requesting these features.   Please attach your sample files.  Any information you have concerning Nikon3.FlashInfo will be appreciated.   Phil has probably documented this on his web-site. 

The tag is known to Exiv2:

```bash
580 rmills@rmillsmm-local:~/gnu/exiv2/team $ taglist ALL | grep Exif.Nikon3.FlashInfo
Nikon3.FlashInfo,168,0x00a8,Nikon3,Exif.Nikon3.FlashInfo,Undefined,"Flash info"
581 rmills@rmillsmm-local:~/gnu/exiv2/team $ 
```

We don’t decode it.  Here’s an example with my D5300:

```bash
584 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ exiv2 -g FlashInfo ~/Stonehenge.jpg 
Exif.Nikon3.FlashInfo  Undefined  49  48 49 48 54 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
585 rmills@rmillsmm-local:~/gnu/github/exiv2/0.27-maintenance $ 
```
So there are 49 bytes which are `"0106\0......\0"`. I know the flash was not used on that photo.  I see there are notes relating to this tag here: https://exiftool.org/TagNames/Nikon.html#FlashInfo0106

I hope the people working on Exiv2 will help you.  Please remember they are all volunteers with jobs, family and other demands on their time.  In anticipation of retirement, I spent 6 months last year writing a book:  **Image Metadata** _and_ **Exiv2 Architecture**  https://exiv2.org/book/index.html  So how to deal with this is documented and I am willing to mentor/assist anybody who volunteers.

If you have good C++ skills, you may enjoy the challenge to bring this feature to life by joining Team Exiv2.  I assure you of my support and assistance.

Robin

--------------------------

Christoph replied and declined my invitation to undertake this work.  He didn't want to open an issue as he knows nothing about git.  So, I offered to open this issue.

### PR DESCRIPTION
**Pull request by postscript-dev**
- Update `Exif.NikonFl7.*` tags
- Add `Exif.NikonFl6` group (with help from ExifTool)
- Remove Exif groups list from the manpage instead of adding new `NikonFl6` group
- Add `NikonFl6` to exiv2 tag webpage documentation
- Update `Exif.Nikon3.FlashMode` with values from ExifTool

Closes #1941.

References:
https://www.exiftool.org/TagNames/Nikon.html
https://github.com/exiftool/exiftool/blob/1e17485cbb372a502e5b9d052d01303db735e6fa/lib/Image/ExifTool/Nikon.pm#L9705
https://github.com/exiftool/exiftool/blob/1e17485cbb372a502e5b9d052d01303db735e6fa/lib/Image/ExifTool/Nikon.pm#L9887

### CODE REVIEW

File: src/nikonmn_int.cpp [hunk 1]
```diff
@@ -804,32 +802,159 @@ constexpr TagDetails nikonFlashFirmware[] = {{0x0000, N_("n/a")},
...
+    {0, N_("Off")},       {1, N_("TTL")},
+    {2, N_("iTTL")},      {3, N_("Auto Aperture")},
+    {4, N_("Automatic")}, {5, N_("GN (distance priority)")},
+    {6, N_("Manual")},    {7, N_("Repeating Flash")},
+    // 8: When used with Group A (and possibly Group B), may have an additional "Quick Wireless Flash Control [A:B]"
+    //    mode that applies to Groups A and B. May only be available with Radio AWL modes?
+    // ?: Possibly includes "iTTL-BL", "i-AA" and "i-A" modes.
 };
 
+constexpr bool flashModeUsesManualScale(const int64_t& mode) {
```
Comment (kevinbackhouse) [inline src/nikonmn_int.cpp:840]:
```suggestion
static constexpr bool flashModeUsesManualScale(const int64_t mode) {
```

File: src/nikonmn_int.cpp [hunk 2]
```diff
@@ -804,32 +802,159 @@ constexpr TagDetails nikonFlashFirmware[] = {{0x0000, N_("n/a")},
...
+//! FlashIlluminationPattern
+constexpr TagDetails nikonFlashIlluminationPat[] = {{0, N_("Illumination Pat.: Standard")},
+                                                    {1, N_("Illumination Pat.: Center-weighted")},
+                                                    {2, N_("Illumination Pat.: Even")}};
+
+//! FlashAdaptors
+constexpr TagDetailsBitmask nikonFlashAdaptors[] = {
+    {0x04, N_("Bounce Flash adaptor")}, {0x10, N_("Wide Flash adaptor")}, {0x20, N_("Nikon Diffusion Dome")}};
+
+void printFlashCompensationValue(std::ostream& os, const unsigned char& value, const bool& manualScale) {
```
Comment (kevinbackhouse) [inline src/nikonmn_int.cpp:877]:
```suggestion
static void printFlashCompensationValue(std::ostream& os, const unsigned char value, const bool manualScale) {
```

File: src/nikonmn_int.cpp [hunk 3]
```diff
@@ -2963,70 +3204,156 @@ std::ostream& Nikon3MakerNote::printExternalFlashData1(std::ostream& os, const V
...
+
+  // Check if using an automated or manual mode
+  auto pos = metadata->findKey(ExifKey("Exif.NikonFl6.ExternalFlashData1"));
+  if (pos == metadata->end() || pos->count() != 1 || pos->typeId() != unsignedByte) {
+    os << "(" << value << ")";
+    os.flags(f);
+    return os;
+  }
+
+  const auto mode = pos->toInt64(0) & 0x0F;
```
Comment (kevinbackhouse) [inline src/nikonmn_int.cpp:3304]: Can you use `toUint32()` here? I think it's cleaner to use unsigned types for bitwise operations like `&`.

File: src/nikonmn_int.cpp [hunk 4]
```diff
@@ -804,32 +802,159 @@ constexpr TagDetails nikonFlashFirmware[] = {{0x0000, N_("n/a")},
...
+    {0, N_("Off")},       {1, N_("TTL")},
+    {2, N_("iTTL")},      {3, N_("Auto Aperture")},
+    {4, N_("Automatic")}, {5, N_("GN (distance priority)")},
+    {6, N_("Manual")},    {7, N_("Repeating Flash")},
+    // 8: When used with Group A (and possibly Group B), may have an additional "Quick Wireless Flash Control [A:B]"
+    //    mode that applies to Groups A and B. May only be available with Radio AWL modes?
+    // ?: Possibly includes "iTTL-BL", "i-AA" and "i-A" modes.
 };
 
+static constexpr bool flashModeUsesManualScale(const int64_t& mode) {
```
Comment (kevinbackhouse) [inline src/nikonmn_int.cpp:840]:
```suggestion
static constexpr bool flashModeUsesManualScale(const int64_t mode) {
```

File: src/nikonmn_int.cpp [hunk 5]
```diff
@@ -804,32 +802,159 @@ constexpr TagDetails nikonFlashFirmware[] = {{0x0000, N_("n/a")},
...
+//! FlashIlluminationPattern
+constexpr TagDetails nikonFlashIlluminationPat[] = {{0, N_("Illumination Pat.: Standard")},
+                                                    {1, N_("Illumination Pat.: Center-weighted")},
+                                                    {2, N_("Illumination Pat.: Even")}};
+
+//! FlashAdaptors
+constexpr TagDetailsBitmask nikonFlashAdaptors[] = {
+    {0x04, N_("Bounce Flash adaptor")}, {0x10, N_("Wide Flash adaptor")}, {0x20, N_("Nikon Diffusion Dome")}};
+
+static void printFlashCompensationValue(std::ostream& os, const unsigned char& value, const bool& manualScale) {
```
Comment (kevinbackhouse) [inline src/nikonmn_int.cpp:877]:
```suggestion
static void printFlashCompensationValue(std::ostream& os, const unsigned char value, const bool manualScale) {
```

File: src/nikonmn_int.cpp (hunk 1 above)
Comment (postscript-dev) [reply src/nikonmn_int.cpp]: Thanks, I have made the change.

File: src/nikonmn_int.cpp (hunk 2 above)
Comment (postscript-dev) [reply src/nikonmn_int.cpp]: Thanks, I have made the change.

File: src/nikonmn_int.cpp (hunk 3 above)
Comment (postscript-dev) [reply src/nikonmn_int.cpp]: Thanks, I have replaced all the `toInt6()` in my code as the tags use unsigned values.

**Comment by postscript-dev** [comment]
@kevinbackhouse:
When I was working on this PR, another change prevented automatic rebasing so I had to do this by hand. After finishing, when I pushed, git requested that I merge the code which I did. I see that a long list of commits have been included in this PR and I am unsure what to do. I was simply planning on squashing the commits when merging but I don't know if this will work. Do I need to do something else?

Also, I think that codeconv is failing due to the way in which merging took place. Exiv2 went through a period where the codeconv percentage produced weird values and I am wondering if merges like this have anything to do with it? This is only a suggestion though, as I don't have any experience with this.

**Comment by kevinbackhouse** [comment]
@postscript-dev: in this situation, I would create a temporary clone of the repo so that I can do potentially destructive things like `git reset`. If something goes wrong then I can just delete the clone and try again. I just gave it a try and I think these commands should work for you:

```bash
git clone https://github.com/Exiv2/exiv2.git exiv2_fix_pr_2274
cd exiv2_fix_pr_2274
git remote add postscript-dev git@github.com:postscript-dev/exiv2.git
git fetch postscript-dev 
git checkout Update_Nikon_flash_tags
git reset --soft main    # This is the dangerous command
git commit -m "Update Nikon flash tags"
```

At this point you should check that everything looks ok and that the tests pass.

Also, this command should return an empty diff, confirming that you didn't accidentally change anything:

```
git diff Update_Nikon_flash_tags postscript-dev/Update_Nikon_flash_tags
```

If everything looks ok, do a force push:

```
git push --force
```

**Comment by postscript-dev** [comment]
@kevinbackhouse:
>  I just gave it a try and I think these commands should work for you:

That worked great, thanks for the help.

**Review by kevinbackhouse — APPROVED** [review]

**Pull request merged**